staging/src/k8s.io/client-go/discovery/cached/memcache_test.go

```diff
@@ -130,3 +133,53 @@ func TestClient(t *testing.T) {
 		t.Errorf("Expected %#v, got %#v", e, a)
 	}
 }
+
+func TestRefreshFails(t *testing.T) {
+	fake := &fakeDiscovery{
+		groupList: &metav1.APIGroupList{
+			Groups: []metav1.APIGroup{{
+				Name: "astronomy",
+				Versions: []metav1.GroupVersionForDiscovery{{
+					GroupVersion: "astronomy/v8beta1",
+					Version:      "v8beta1",
+				}},
+			}},
+		},
+		resourceMap: map[string]*metav1.APIResourceList{
+			"astronomy/v8beta1": {
+				GroupVersion: "astronomy/v8beta1",
+				APIResources: []metav1.APIResource{{
+					Name:         "dwarfplanets",
+					SingularName: "dwarfplanet",
+					Namespaced:   true,
+					Kind:         "DwarfPlanet",
+					ShortNames:   []string{"dp"},
+				}},
+			},
+		},
+		err: errors.New("some error"),
+	}
+
+	c := NewMemCacheClient(fake)
+	if c.Fresh() {
+		t.Errorf("Expected not fresh.")
+	}
+	g, err := c.ServerGroups()
+	if err == nil {
+		t.Errorf("Expected error")
+	}
+	if c.Fresh() {
+		t.Errorf("Expected not fresh.")
+	}
+	fake.err = nil
+	g, err = c.ServerGroups()
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+	if e, a := fake.groupList, g; !reflect.DeepEqual(e, a) {
+		t.Errorf("Expected %#v, got %#v", e, a)
+	}
+	if !c.Fresh() {
+		t.Errorf("Expected fresh.")
+	}
+}
```

Comment:
Can you add a test for partial failure, i.e., discovery fails at a specific group?

Reply:
Added